File: trunk/xml-only/src/widgets/DatabaseConnectionDialog.cpp
#include "DatabaseConnectionDialog.h"

#include <QLayout>
#include <QLineEdit>
#include <QSpinBox>
#include <QCheckBox>
#include <QDialogButtonBox>
#include <QGroupBox>
#include <QLabel>
#include <QApplication>

#include "HostEdit.h"
#include "LoginEdit.h"
#include "PasswordEdit.h"
#include "DatabaseConnectionSettings.h"
#include "Settings.h"
#include "ProfileLogger.h"

DatabaseConnectionDialog::DatabaseConnectionDialog(QWidget* p)
  : QDialog(p) {
  setLayout(new QVBoxLayout(this));
  QGroupBox* b = new QGroupBox(tr("Database Connection"), this);
  _bb = new QDialogButtonBox(QDialogButtonBox::Ok | QDialogButtonBox::Cancel,
			     Qt::Horizontal,
			     this);
  QGridLayout* bl = new QGridLayout(b);
  int r = 0;
  int lC = 0;
  int wC = 1;

  QLabel* hostL = new QLabel(tr("&Host"), b);
  _hostW = new HostEdit(b);
  hostL->setBuddy(_hostW);
  QLabel* portsL = new QLabel(tr("&Port"), b);
  _portsW = new QSpinBox(b);
  _portsW->setRange(1, 999999);
  portsL->setBuddy(_portsW);
  QLabel* dbL = new QLabel(tr("&Database"), b);
  _dbW = new QLineEdit(b);
  dbL->setBuddy(_dbW);
  QLabel* userL = new QLabel(tr("&User"), b);
  _userW = new LoginEdit(b);
  userL->setBuddy(_userW);
  QLabel* passL = new QLabel(tr("&Password"), b);
  _passW = new PasswordEdit(b);
  passL->setBuddy(_passW);
  _dropSchemaW = new QCheckBox(tr("Drop Schema"), b);
  _createSchemaW = new QCheckBox(tr("Create Schema"), b);
  _insertTemplateDataW = new QCheckBox(tr("Insert Template Data"), b);

  bl->addWidget(hostL, r, lC);
  bl->addWidget(_hostW, r, wC);
  r++;
  bl->addWidget(portsL, r, lC);
  bl->addWidget(_portsW, r, wC);
  r++;
  bl->addWidget(dbL, r, lC);
  bl->addWidget(_dbW, r, wC);
  r++;
  bl->addWidget(userL, r, lC);
  bl->addWidget(_userW, r, wC);
  r++;
  bl->addWidget(passL, r, lC);
  bl->addWidget(_passW, r, wC);
  r++;
  bl->addWidget(_dropSchemaW, r, wC);
  r++;
  bl->addWidget(_createSchemaW, r, wC);
  r++;
  bl->addWidget(_insertTemplateDataW, r, wC);

  layout()->addWidget(b);
  layout()->addWidget(_bb);

  connect(_hostW, SIGNAL(textChanged(const QString&)), this, SLOT(slotHostChanged(const QString&)));
  connect(_portsW, SIGNAL(valueChanged(int)), this, SLOT(slotPortChanged(int)));
  connect(_dbW, SIGNAL(textChanged(const QString&)), this, SLOT(slotDatabaseChanged(const QString&)));
  connect(_userW, SIGNAL(textChanged(const QString&)), this, SLOT(slotUserChanged(const QString&)));
  connect(_passW, SIGNAL(textChanged(const QString&)), this, SLOT(slotPasswordChanged(const QString&)));
  connect(_dropSchemaW, SIGNAL(toggled(bool)), this, SLOT(slotDropSchemaToggled(bool)));
  connect(_createSchemaW, SIGNAL(toggled(bool)), this, SLOT(slotCreateSchemaToggled(bool)));
  connect(_insertTemplateDataW, SIGNAL(toggled(bool)), this, SLOT(slotInsertTemplateDataToggled(bool)));
  connect(_bb, SIGNAL(accepted()), this, SLOT(accept()));
  connect(_bb, SIGNAL(rejected()), this, SLOT(reject()));

  _s = (static_cast<ProfileLogger*>(QApplication::instance()))->getSettings()->getDatabaseConnectionSettings();
  _hostW->setText(_s.getHost());
  _portsW->setValue(_s.getPort());
  _dbW->setText(_s.getDatabase());
  _userW->setText(_s.getUser());
  _passW->clear();
  if (_s.getDropSchema()) {
    _dropSchemaW->setCheckState(Qt::Checked);
  }
  if (_s.getCreateSchema()) {
    _createSchemaW->setCheckState(Qt::Checked);
  }
  if (_s.getInsertTemplateData()) {
    _insertTemplateDataW->setCheckState(Qt::Checked);
  }
}

void DatabaseConnectionDialog::slotHostChanged(const QString& s) {
  _s.setHost(s);
}

void DatabaseConnectionDialog::slotPortChanged(int p) {
  _s.setPort(p);
}

void DatabaseConnectionDialog::slotDatabaseChanged(const QString& s) {
  _s.setDatabase(s);
}

void DatabaseConnectionDialog::slotUserChanged(const QString& s) {
  _s.setUser(s);
}

void DatabaseConnectionDialog::slotPasswordChanged(const QString& s) {
  _s.setPassword(s);
  _pass = s;
}

void DatabaseConnectionDialog::slotDropSchemaToggled(bool b) {
  _s.setDropSchema(b);
}

void DatabaseConnectionDialog::slotCreateSchemaToggled(bool b) {
  _s.setCreateSchema(b);
}

void DatabaseConnectionDialog::slotInsertTemplateDataToggled(bool b) {
  _s.setInsertTemplateData(b);
}

DatabaseConnectionDialog::~DatabaseConnectionDialog() {
}

void DatabaseConnectionDialog::accept() {
  (static_cast<ProfileLogger*>(QApplication::instance()))->getSettings()->setDatabaseConnectionSettings(_s);
  done(QDialog::Accepted);
}

File: branches/experimental/py/Gui/Dialogs/DatabaseConnectionDialog.py
from PyQt4.QtGui import *
from PyQt4.QtCore import *

class DatabaseConnectionDialog(QDialog):
    def __init__(self, connectionData, parent):
        super(DatabaseConnectionDialog, self).__init__(parent)
        self.cd = connectionData

        self.setLayout(QVBoxLayout(self))

        self.gb = QGroupBox(self)
        self.gb.setTitle(self.tr('Database Connection'))
        self.bb = QDialogButtonBox(QDialogButtonBox.Ok | QDialogButtonBox.Cancel,
                                   Qt.Horizontal, self)
        self.bb.accepted.connect(self.accept)
        self.bb.rejected.connect(self.reject)

        self.layout().addWidget(self.gb)
        self.layout().addWidget(self.bb)

        self.setupWidgets()
        self.hostW.setText(str(self.cd.host))
        self.portW.setText(str(self.cd.port))
        self.dbW.setText(str(self.cd.dbName))
        self.schemaW.setText(str(self.cd.schema))
        self.loginW.setText(str(self.cd.user))
        self.passwordW.setText(str(self.cd.password))
        if self.cd.dropSchema:
            self.dropSchemaW.setCheckState(Qt.Checked)
        if self.cd.createSchema:
            self.createSchemaW.setCheckState(Qt.Checked)
        if self.cd.insertTemplateData:
            self.insertTemplateDataW.setCheckState(Qt.Checked)
        self.passwordW.setFocus()

    def setupWidgets(self):
        self.gb.setLayout(QGridLayout(self.gb))

        lC = 0;
        wC = 1
        r = 0

        self.hostL = QLabel(self.tr('&Host'), self.gb)
        self.hostW = QLineEdit(self.gb)
        self.hostL.setBuddy(self.hostW)
        self.gb.layout().addWidget(self.hostL, r, lC)
        self.gb.layout().addWidget(self.hostW, r, wC);
        self.hostW.textChanged.connect(self.onHostChange)
        r += 1

        self.portL = QLabel(self.tr('&Port'), self.gb)
        self.portW = QLineEdit(self.gb)
        self.gb.layout().addWidget(self.portL, r, lC)
        self.portL.setBuddy(self.portW)
        self.gb.layout().addWidget(self.portW, r, wC);
        self.portW.textChanged.connect(self.onPortChange)
        r += 1

        self.dbL = QLabel(self.tr('&Database'), self.gb)
        self.dbW = QLineEdit(self.gb)
        self.dbL.setBuddy(self.dbW)
        self.gb.layout().addWidget(self.dbL, r, lC)
        self.gb.layout().addWidget(self.dbW, r, wC);
        self.dbW.textChanged.connect(self.onDatabaseChange)
        r += 1

        self.schemaL = QLabel(self.tr('&Schema'), self.gb)
        self.schemaW = QLineEdit(self.gb)
        self.schemaL.setBuddy(self.schemaW)
        self.gb.layout().addWidget(self.schemaL, r, lC)
        self.gb.layout().addWidget(self.schemaW, r, wC);
        self.schemaW.textChanged.connect(self.onSchemaChange)
        r += 1

        self.loginL = QLabel(self.tr('&Login'), self.gb)
        self.loginW = QLineEdit(self.gb)
        self.loginL.setBuddy(self.loginW)
        self.gb.layout().addWidget(self.loginL, r, lC)
        self.gb.layout().addWidget(self.loginW, r, wC);
        self.loginW.textChanged.connect(self.onLoginChange)
        r += 1

        self.passwordL = QLabel(self.tr('P&assword'), self.gb)
        self.passwordW = QLineEdit(self.gb)
        self.passwordW.setEchoMode(QLineEdit.Password)
        self.passwordL.setBuddy(self.passwordW)
        self.gb.layout().addWidget(self.passwordL, r, lC)
        self.gb.layout().addWidget(self.passwordW, r, wC);
        self.passwordW.textChanged.connect(self.onPasswordChange)
        r += 1

        self.dropSchemaW = QCheckBox(self.tr("Drop Schema"), self.gb)
        self.gb.layout().addWidget(self.dropSchemaW, r, wC);
        self.dropSchemaW.toggled.connect(self.onDropSchemaToggle)
        r += 1

        self.createSchemaW = QCheckBox(self.tr("Create Schema"), self.gb)
        self.gb.layout().addWidget(self.createSchemaW, r, wC);
        self.createSchemaW.toggled.connect(self.onCreateSchemaToggle)
        r += 1

        self.insertTemplateDataW = QCheckBox(self.tr("Insert Template Data"), self.gb)
        self.gb.layout().addWidget(self.insertTemplateDataW, r, wC)
        self.insertTemplateDataW.toggled.connect(self.onInsertTemplateDataToggle)

    def onHostChange(self, txt):
        self.cd.host = str(txt)
    def onPortChange(self, port):
        self.cd.port = str(port)
    def onDatabaseChange(self, db):
        self.cd.dbName = str(db)
    def onSchemaChange(self, s):
        self.cd.schema = s
    def onLoginChange(self, l):
        self.cd.user = str(l)
    def onPasswordChange(self, p):
        self.cd.password = str(p)
    def onDropSchemaToggle(self, t):
        self.cd.dropSchema = t
    def onCreateSchemaToggle(self, t):
        self.cd.createSchema = t
    def onInsertTemplateDataToggle(self, t):
        self.cd.insertTemplateData = t
